panic: runtime error: index out of range

I have a zk cluster with 5 nodes. I deployed zookeeper_exporter to all nodes, using command 
```
sudo nohup path_to_zookeeper_exporter &
```
with no additional arguements.
I frequently encounter the following bug(exporter crashed):
```
time="2017-08-24T16:09:29+08:00" level=info msg="Starting zookeeper_exporter"
time="2017-08-24T16:09:29+08:00" level=info msg="Starting metric http endpoint on :9141"
time="2017-08-24T16:09:34+08:00" level=info msg="Fetching metrics from Zookeeper"
panic: runtime error: index out of range

goroutine 31 [running]:
main.(*zookeeperCollector).Collect(0xc420011ec0, 0xc420148360)
        /home/qunheadmin/go/src/github.com/carlpett/zookeeper_exporter/zookeeper.go:154 +0x59d
github.com/carlpett/zookeeper_exporter/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func2(0xc4200fa400, 0xc420148360, 0x8978e0, 0xc420011ec0)
        /home/qunheadmin/go/src/github.com/carlpett/zookeeper_exporter/vendor/github.com/prometheus/client_golang/prometheus/registry.go:433 +0x61
created by github.com/carlpett/zookeeper_exporter/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather
        /home/qunheadmin/go/src/github.com/carlpett/zookeeper_exporter/vendor/github.com/prometheus/client_golang/prometheus/registry.go:434 +0x2ec
time="2017-08-24T16:15:29+08:00" level=info msg="Starting zookeeper_exporter"
time="2017-08-24T16:15:29+08:00" level=info msg="Starting metric http endpoint on :9141"
time="2017-08-24T16:15:34+08:00" level=info msg="Fetching metrics from Zookeeper"
panic: runtime error: index out of range

goroutine 52 [running]:
main.(*zookeeperCollector).Collect(0xc4200fa7c0, 0xc42005e9c0)
        /home/qunheadmin/go/src/github.com/carlpett/zookeeper_exporter/zookeeper.go:154 +0x59d
github.com/carlpett/zookeeper_exporter/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func2(0xc420011b00, 0xc42005e9c0, 0x8978e0, 0xc4200fa7c0)
        /home/qunheadmin/go/src/github.com/carlpett/zookeeper_exporter/vendor/github.com/prometheus/client_golang/prometheus/registry.go:433 +0x61
created by github.com/carlpett/zookeeper_exporter/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).Gather
        /home/qunheadmin/go/src/github.com/carlpett/zookeeper_exporter/vendor/github.com/prometheus/client_golang/prometheus/registry.go:434 +0x2ec
```
Sometimes this bug occurs right after I start exporter. Sometimes it occurs after a long period(more than one day).

Any thoughts about this? Thanks.

--- a/zookeeper.go
+++ b/zookeeper.go
@@ -6,6 +6,7 @@ import (
 	"net"
 	"strconv"
 	"strings"
+	"time"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/prometheus/client_golang/prometheus"
@@ -149,8 +150,14 @@ func (c *zookeeperCollector) Collect(ch chan<- prometheus.Metric) {
 	}
 
 	data = strings.TrimSpace(data)
+	status := 1.0
 	for _, line := range strings.Split(data, "\n") {
 		parts := strings.Split(line, "\t")
+		if len(parts) != 2 {
+			log.WithFields(log.Fields{"data": line}).Warn("Unexpected format of returned data, expected tab-separated key/value.")
+			status = 0
+			continue
+		}
 		label, value := parts[0], parts[1]
 		metric, ok := c.metrics[label]
 		if ok {
@@ -162,7 +169,7 @@ func (c *zookeeperCollector) Collect(ch chan<- prometheus.Metric) {
 			}
 		}
 	}
-	ch <- prometheus.MustNewConstMetric(c.upIndicator, prometheus.GaugeValue, 1)
+	ch <- prometheus.MustNewConstMetric(c.upIndicator, prometheus.GaugeValue, status)
 
 	if *resetOnScrape {
 		resetStatistics()
@@ -176,6 +183,10 @@ func resetStatistics() {
 	}
 }
 
+const (
+	timeoutSeconds = 5
+)
+
 func sendZkCommand(fourLetterWord string) (string, bool) {
 	log.Debugf("Connecting to Zookeeper at %s", *zookeeperAddr)
 
@@ -186,8 +197,18 @@ func sendZkCommand(fourLetterWord string) (string, bool) {
 	}
 	defer conn.Close()
 
+	err = conn.SetDeadline(time.Now().Add(timeoutSeconds * time.Second))
+	if err != nil {
+		log.WithFields(log.Fields{"error": err}).Error("Failed to set timeout on Zookeeper connection")
+		return "", false
+	}
+
 	log.WithFields(log.Fields{"command": fourLetterWord}).Debug("Sending four letter word")
-	conn.Write([]byte(fourLetterWord))
+	_, err = conn.Write([]byte(fourLetterWord))
+	if err != nil {
+		log.WithFields(log.Fields{"error": err}).Error("Error sending command to Zookeeper")
+		return "", false
+	}
 	scanner := bufio.NewScanner(conn)
 
 	buffer := bytes.Buffer{}

